### ISSUE TITLE
Reusing a moved `tcp::socket` in `async_accept` throws `bad_excecutor`

### ISSUE THREAD

**Issue by laudrup** [chriskohlhoff/asio#790]
Using a moved from socket as an argument to `tcp::acceptor::async_accept` causes a `bad_executor` exception to be thrown.

This is the smallest reproducable example I could come up with:
```C++
#include <boost/asio.hpp>

#include <array>
#include <string>
#include <thread>

using namespace boost::asio;

class move_socket_server {
public:
    move_socket_server(io_context& context)
        : context_(context)
        , acceptor_(context_, ip::tcp::endpoint(ip::tcp::v4(), 6969))
        , socket_(context_)
        , accept_socket_(context_) {
        do_accept();
    }

    void do_accept() {
        acceptor_.async_accept(accept_socket_, [this](boost::system::error_code) {
            if (socket_.is_open()) {
                static const std::string busy_reply = "Server busy\n";
                async_write(accept_socket_, buffer(busy_reply), [this](const boost::system::error_code, size_t) {
                    do_accept();
                });
            } else {
                static const std::string hello_reply = "Hello from server\n";
                // accept_socket_ = ip::tcp::socket(context_); // <- Uncomment this as a workaround
                socket_ = std::move(accept_socket_);
                // bad_executor thrown on second accept
                async_write(socket_, buffer(hello_reply), [this](boost::system::error_code, size_t) {
                    socket_.close();
                });
                do_accept();
            }
        });
    }

private:
    io_context& context_;
    ip::tcp::acceptor acceptor_;
    ip::tcp::socket socket_;
    ip::tcp::socket accept_socket_;
};

int main() {
    io_context context;
    move_socket_server server(context);
    std::thread t([&context]() {
        context.run();
    });

    for (int i=0; i<2; ++i) {
        ip::tcp::socket socket(context);
        std::array<char, 1024> buf;
        boost::system::error_code ec;
        socket.connect(ip::tcp::endpoint(ip::make_address("127.0.0.1"), 6969), ec);
        socket.read_some(buffer(buf), ec);
    }

    context.stop();
    t.join();
    return 0;
}
```
This works fine in `boost::asio` 1.73 and I cannot see why it shouldn't work. As noted in the example, a workaround is to reinitialize the moved from socket, but as the documentation states, a moved from socket should be in the exact same state as if it was constructed with an executor, so I assume this is a bug.

Seems like this could be related to [issue 381](https://github.com/chriskohlhoff/asio/issues/381)

**Comment by kashirin-alex** [comment]
There some wrong things in the code:
1. the move socket `socket_ = std::move(accept_socket_);` is not thread-safe
2. if and accept_socket_ was moved (as-should), with the second move it be an empty instance (at least expected to be, incl. the executor)

-- May be, It is a result of matching the Move behavior. (eg. a pointer in a moved instance assigned with a nullptr)

**Comment by kashirin-alex** [comment]
some of the movers involve:
https://github.com/chriskohlhoff/asio/blob/b2399796920dcdf73a2b306ab51619e660c72bda/asio/include/asio/detail/io_object_impl.hpp#L98
https://github.com/chriskohlhoff/asio/blob/0355fc59807198034780c814c799ed353ffe960f/asio/include/asio/io_context.hpp#L694

**Comment by laudrup** [comment]
Thanks a lot for your quick reply!

> There are some wrong things in the code:
> 
> 1. the move socket `socket_ = std::move(accept_socket_);` is not thread-safe

While I usually avoid using threads if at all possible, in this case it just seemed like the simplest way to write an example and I cannot see how the (missing) thread safety of the move constructor is relevant here . The same issue exists if no threads are involved and only the server is started but more than one client connects from a separate process.

> 2. if and accept_socket_ was moved (as-should), with the second move it be an empty instance (at least expected to be, incl. the executor)
> 
> -- May be, It is a result of matching the Move behavior. (eg. a pointer in a moved instance assigned with a nullptr)

That sounds a lot more like the problem, but the [documentation for the sockets move constructor states](https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/basic_stream_socket/basic_stream_socket/overload9.html):

> Following the move, the moved-from object is in the same state as if constructed using the basic_stream_socket(const executor_type&) constructor.

Doesn't that mean that it should be possible to use a moved-from socket in eg. a call to `async_accept` like in the code I posted? At least according to the documentation?

It's very likely I've misunderstood something, it just looked like a potential bug in asio code to me and while the code where I've experienced this should be completely rewritten anyway, I thought it better not to just ignore it.

Thanks for your time.

**Comment by kashirin-alex** [comment]
In this case, including the [assign-operator](https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/basic_stream_socket/operator_eq_/overload1.html), I can't see the reason except that something might have gone wrong (as you state previous version was ok with using the move socket).

as to the `executor_.~executor_type();`  is no wrong (overlooked by a line)
https://github.com/chriskohlhoff/asio/blob/b2399796920dcdf73a2b306ab51619e660c72bda/asio/include/asio/detail/io_object_impl.hpp#L98

-- I use the returned socket by the `async_accept` and I hadn't have a try with moving a socket instance in-partial sense.
https://github.com/kashirin-alex/swc-db/blob/9e44c99fbbb53bbdaedeed4d58a24364e560bbe0/src/cc/lib/swcdb/core/comm/SerializedServer.cc#L80

**Comment by laudrup** [comment]
@kashirin-alex So do you agree that this looks like a regression, or perhaps the documentation for `tcp::socket`s move operator that needs to be changed?

**Comment by kashirin-alex** [comment]
sorry, I don't know the plans(changes & docs correlation) for the result of the socket move

**Comment by laudrup** [comment]
@kashirin-alex That's fine. I would still consider this to be a regression/bug, so I'll leave this issue open.

Thanks a lot for all your input.

**Comment by laudrup** [comment]
This has been fixed in the latest release of Boost (1.76 - Asio 1.18.2).

I haven't test specifically, but I assume [this commit](https://github.com/chriskohlhoff/asio/commit/b5a8065ecc018b521ed800475b59b8423f3c8564) fixes it.

Thanks a lot.
